### DISCUSSION TITLE
Enable/Disable Dynamic Powerlimiter via HTTP

### DISCUSSION THREAD

**Discussion by ele-dev** [hoylabs/OpenDTU-OnBattery#469] [Q&A] [answered]
I have a simple question. I want to en-/disable the dynamic power limiter via an HTTP request. I read the wiki about the HTTP API but couldn't figure out a working HTTP request that does the job.

Can you tell me if it even can be done via HTTP?
What exact URL is needed and what parameters?

**Comment by spcqike** [comment, marked as answer]
as the WebUI uses the API, it can be done. however, you need to create a request, that imitates the webui "save" button

e.g. 
> curl -u "user:password" http://ip/api/powerlimiter/config -d 'data={"enabled":"true","lower_power_limit":"10","inverter_id":"ID 00","inverter_channel_id":"CH 0","target_power_consumption":"-20","target_power_consumption_hysteresis":"10"}'

note, that the parameters are not complete. you must add all parameters, you want to set. the 6 paremeters i wrote are the bare minimum. if one is missing, the response will be  "Values are missing!".

all parameters can be found here 
https://github.com/helgeerbe/OpenDTU-OnBattery/blob/4ee49a6ecb90a800e57db7554a88b53cb0bf55af/src/WebApi_powerlimiter.cpp#L126C1-L151C1

in a quick test: if you let parameters out, they will be "0". so keep that in mind for your upper Powerlimit, SoC states etc.


edit, just checked and played around. in my case it works with

> {"enabled":true,"verbose_logging":false,"solar_passthrough_enabled":true,"solar_passthrough_losses":0,"battery_drain_strategy":0,"is_inverter_behind_powermeter":true,"inverter_id":0,"inverter_channel_id":0,"target_power_consumption":-20,"target_power_consumption_hysteresis":10,"lower_power_limit":10,"upper_power_limit":600,"battery_soc_start_threshold":0,"battery_soc_stop_threshold":0,"voltage_start_threshold":0,"voltage_stop_threshold":0,"voltage_load_correction_factor":0,"inverter_restart_hour":0,"full_solar_passthrough_soc":0,"full_solar_passthrough_start_voltage":0,"full_solar_passthrough_stop_voltage":0}

and 

> {"enabled":false,"verbose_logging":false,"solar_passthrough_enabled":true,"solar_passthrough_losses":0,"battery_drain_strategy":0,"is_inverter_behind_powermeter":true,"inverter_id":0,"inverter_channel_id":0,"target_power_consumption":-20,"target_power_consumption_hysteresis":10,"lower_power_limit":10,"upper_power_limit":600,"battery_soc_start_threshold":0,"battery_soc_stop_threshold":0,"voltage_start_threshold":0,"voltage_stop_threshold":0,"voltage_load_correction_factor":0,"inverter_restart_hour":0,"full_solar_passthrough_soc":0,"full_solar_passthrough_start_voltage":0,"full_solar_passthrough_stop_voltage":0}

to get your payload easily, you can set it up in webui, query it with

> curl -u "user:password" http://ip/api/powerlimiter/config

and just change "true" and "false" accordingly

**Reply by ele-dev** [reply]
Thanks for the reply. I will try it out

**Reply by spcqike** [reply]
also keep in mind, that disableing the DPL also stopps the inverter. so you must restart it afterwards.

**Reply by ele-dev** [reply]
Yes but that's intentional. I use an external logic for AC charging during the day and therefore wanted a safe way to strictly control when charging (day) and discharging (night) of the battery is done to avoid any conflicts. It's basically a state machine.

**Reply by stanpete78** [reply]
That's interesting, thank you. But if you'd build some kind of automation with it, would changing/saving those values result in a degradation of the Hoymiles' EPROM or any other component that maybe suffers from frequent saves?

**Reply by ele-dev** [reply]
I don't think so, to my knowledge the dynamic power limiter does not write to the eeprom on the hoymiles because nothing needs to be stored permanently on the inverter.  I'm pretty sure dpl uses the temporary power limit instructions that are lost after a restart of the inverter.

**Reply by stanpete78** [reply]
If I (de)activate the DPL, the "saves" counter on the info/system page goes up by one. This makes me a bit nervous.

**Reply by schlimmchen** [reply]
Hm? That counts the amount of times the config was written to the ESP32 flash.

**Reply by stanpete78** [reply]
So I could write those values every minute and it would'nt be a bad thing? Ok, good to know.

**Reply by schlimmchen** [reply]
> I'm pretty sure dpl uses the temporary power limit instructions that are lost after a restart of the inverter.

Yes.

**Reply by schlimmchen** [reply]
> So I could write those values every minute and it would'nt be a bad thing?

Well, that's 1440 writes a day. I dont't know anything about the implementation of the flash filesystem on the ESP32, but if its wear leveling is bad, this might lead to write errors "soon". Let's say within a year? Worst case: within 10 days.

It's a little excessive. Try it out and replace the ESP32 if it breaks. It's cheap.

**Reply by stanpete78** [reply]
I found this: "ESP32s use [NOR flash storage](https://en.wikipedia.org/wiki/Flash_memory#NOR_memories), which is usually rated for between 10,000 to 100,000 write cycles (100,000 seems to be the standard these days). Flash can't write single bytes; instead of writes a "page" of bytes, which I believe is 256 bytes. So each 256 byte page is rated for at least 100,000 cycles. When a device is rated for 100,000 cycles it's likely to be usable for at least 10 times that, but the manufacturer is not going to make any promises beyond the 100,000."

**Reply by schlimmchen** [reply]
So, even with no wear leveling at all (very unlikely), this should be fine for at least 100 days
